### PR TITLE
[#4] Setup LogIn and SignUp

### LINKED ISSUE #4: Setup Login & Sign up

### PR DESCRIPTION
**Pull request by jferreira93**
Resolves #4

### CODE REVIEW

File: spec/factories/users.rb [hunk 1]
```diff
@@ -18,11 +17,13 @@
...
 FactoryBot.define do
   factory :user do
-    password 'secret_pass'
     name 'Random Name'
     photo 'random_blob'
     phone '0293849393'
     role :no_admin
+    token 'token'
+    refresh_token 'refresh_token'
+    oauth_expires_at Time.new(2020)
```
Comment (lucaleivaloop) [inline spec/factories/users.rb]: Why `2020` is being used here?

File: db/migrate/20181008145233_change_oauth_expires_at_to_be_datetime_in_users.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+# This migration change oauth_expires_at datatype
+class ChangeOauthExpiresAtToBeDatetimeInUsers < ActiveRecord::Migration[5.2]
+  def up
+    change_column :users, :oauth_expires_at, '
+      timestamp USING CAST(oauth_expires_at AS timestamp)
+    '
```
Comment (lucaleivaloop) [inline db/migrate/20181008145233_change_oauth_expires_at_to_be_datetime_in_users.rb]: Wouldn't this be better in one line after the last comma? Ex: 
```ruby
change_column :users, :oauth_expires_at, 
'timestamp USING CAST(oauth_expires_at AS timestamp)'
```

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -1,2 +1,10 @@
+# frozen_string_literal: true
+
+# This is the application controller
 class ApplicationController < ActionController::Base
+  protect_from_forgery with: :exception
+
+  def index
```
Comment (lucaleivaloop) [inline app/controllers/application_controller.rb]: This might be better as
```suggestion
  def index; end
```

File: spec/models/user_spec.rb [hunk 4]
```diff
@@ -16,7 +16,7 @@
 #
 
 require 'rails_helper'
-
+# rubocop:disable Metrics/BlockLength
```
Comment (mconiglio) [inline spec/models/user_spec.rb]: This could happen in a lot of specs files. What about configuring this rule to be excluded on spec folder?
See this: https://docs.rubocop.org/en/latest/configuration/#includingexcluding-files
And this: https://github.com/rubocop-hq/rubocop/issues/3694#issuecomment-257749697

File: spec/models/user_spec.rb [hunk 5]
```diff
@@ -25,16 +25,112 @@
...
+        },
+        info: {
+          email: 'some@email.com',
+          name: 'some_name'
+        }
+      }
+    end
+
+    context 'when user does not exist' do
+      it 'should return new user' do
```
Comment (mconiglio) [inline spec/models/user_spec.rb]: I don't think this is being tested correctly. maybe you could check that the User.count is incremented by 1 when the subject is executed `it { expect{ subject }.to change(...).by(1) }`

File: spec/models/user_spec.rb [hunk 6]
```diff
@@ -25,16 +25,112 @@
...
+          expires_at: '01/01/2020'
+        },
+        info: {
+          email: 'some@email.com',
+          name: 'some_name'
+        }
+      }
+    end
+
+    context 'when user does not exist' do
```
Comment (mconiglio) [inline spec/models/user_spec.rb]: also check that the result is persisted and that it's an instance of `User` class

File: spec/models/user_spec.rb [hunk 7]
```diff
@@ -25,16 +25,112 @@
...
+            email: 'some@email.com',
+            name: 'some_name'
+          }
+        }
+      end
+
+      it 'should update attributes' do
+        expect(subject.uuid).to eq(user.uuid)
+        expect(subject.token).to eq('some_token')
+        expect(subject.refresh_token)
```
Comment (mconiglio) [inline spec/models/user_spec.rb]: why are you splitting this line in two? it's not longer than 80 chars

File: spec/models/user_spec.rb [hunk 8]
```diff
@@ -25,16 +25,112 @@
...
+          },
+          info: {
+            email: 'some@email.com',
+            name: 'some_name'
+          }
+        }
+      end
+
+      it 'should update attributes' do
+        expect(subject.uuid).to eq(user.uuid)
```
Comment (mconiglio) [inline spec/models/user_spec.rb]: every time you call "subject" you are calling the `update_or_create` method. you could use `result = subject` and check for values on the result object like `result.uuid`, etc.

File: spec/models/user_spec.rb [hunk 9]
```diff
@@ -25,16 +25,112 @@
...
+        expect(subject.token).to eq('some_token')
+        expect(subject.refresh_token)
+          .to eq('some_refresh_token')
+        expect(subject.oauth_expires_at)
+          .to eq('01/01/2020')
+        expect(subject.email).to eq('some@email.com')
+        expect(subject.name).to eq('some_name')
+        subject
```
Comment (mconiglio) [inline spec/models/user_spec.rb]: this line is not necessary

File: spec/models/user_spec.rb [hunk 10]
```diff
@@ -25,16 +25,112 @@
...
+        info: {
+          email: 'some@email.com',
+          name: 'some_name'
+        }
+      }
+    end
+
+    context 'receive an auth' do
+      it 'normalize auth response' do
+        subject
```
Comment (mconiglio) [inline spec/models/user_spec.rb]: remove this line. `is_expected` already calls the subject implicitly

File: spec/models/user_spec.rb [hunk 11]
```diff
@@ -25,16 +25,112 @@
...
+          expires_at: '01/01/2020'
+        },
+        info: {
+          email: 'some@email.com',
+          name: 'some_name'
+        }
+      }
+    end
+
+    context 'receive an auth' do
```
Comment (mconiglio) [inline spec/models/user_spec.rb]: why are you creating this context? it's not necessary

File: spec/factories/users.rb [hunk 12]
```diff
@@ -18,11 +17,13 @@
...
 FactoryBot.define do
   factory :user do
-    password 'secret_pass'
     name 'Random Name'
     photo 'random_blob'
     phone '0293849393'
     role :no_admin
+    token 'token'
+    refresh_token 'refresh_token'
+    oauth_expires_at (Time.now + 80.days)
```
Comment (mconiglio) [inline spec/factories/users.rb]: I think you should use block for this, `{ Time.now + 80.days }`

File: spec/controllers/session_controller_spec.rb [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+
+      it 'should redirect to root path' do
+        subject
+        expect(response).to redirect_to(root_path)
+      end
+    end
+  end
+
+  describe '#destroy' do
+    context 'correct flow' do
```
Comment (mconiglio) [inline spec/controllers/session_controller_spec.rb]: Maybe `'when the session is active'` would be more appropriate

File: spec/controllers/session_controller_spec.rb [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+      end
+
+      it 'should redirect to root path' do
+        subject
+        expect(response).to redirect_to(root_path)
+      end
+    end
+  end
+
+  describe '#destroy' do
```
Comment (mconiglio) [inline spec/controllers/session_controller_spec.rb]: test cases for when session is inactive?

File: spec/models/user_spec.rb (hunk 4 above)
Comment (mconiglio) [inline spec/models/user_spec.rb]: same comment about disabling this rule for specs

File: config/routes.rb [hunk 15]
```diff
@@ -1,5 +1,12 @@
 Rails.application.routes.draw do
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+
+  get "/login", to: redirect("/auth/google_oauth2")
```
Comment (mconiglio) [inline config/routes.rb]: use single quotes

File: spec/models/user_spec.rb (hunk 4 above)
Comment (jferreira93) [reply spec/models/user_spec.rb]: The configuration .rubocop.yml doesnot exist yet

File: spec/models/user_spec.rb (hunk 10 above)
Comment (mconiglio) [reply spec/models/user_spec.rb]: `subject` line can be removed now

File: spec/models/user_spec.rb (hunk 4 above)
Comment (mconiglio) [reply spec/models/user_spec.rb]: you can create it haha

File: app/views/layouts/_nav_layout.html.erb [hunk 16]
```diff
@@ -1,55 +1,55 @@
...
-      </div>
-      <div class="top-nav-button" id="projects-button">
-        <i class="fas fa-clock"></i>
-      </div>
-      <div class="top-nav-button" id="projects-button">
-        <i class="fas fa-hand-peace "></i>
-      </div>
-    </div>
-  </div>
+ <div class="mdl-layout__header-row">
```
Comment (mconiglio) [inline app/views/layouts/_nav_layout.html.erb]: why did you change this? the indentation was correct

File: spec/models/user_spec.rb [hunk 17]
```diff
@@ -25,16 +24,111 @@
...
+      it 'normalize auth response' do
+        subject
```
Comment (mconiglio) [inline spec/models/user_spec.rb]: remove

File: app/views/layouts/_nav_layout.html.erb [hunk 18]
```diff
@@ -1,55 +1,55 @@
...
-        <i class="fas fa-hand-peace "></i>
-      </div>
-    </div>
-  </div>
+<header class="mdl-layout__header">
+ <div class="mdl-layout__header-row">
+   <span class="message">What have you done?</span>
+   <div class="layout-spacer">
+     <div class="top-nav-button" id="projects-button">
+       <i class="fas fa-folder"></i></i>
```
Comment (mconiglio) [inline app/views/layouts/_nav_layout.html.erb]: you are closing the tag two times
